File: src/twicc/core/text_filter.py
"""Shared text-filter grammar for frontend-equivalent backend filtering."""


def match_subsequence(query: str, text: str) -> bool:
    """Return whether every query character occurs in text, in order."""
    lower_query = query.lower()
    lower_text = text.lower()
    query_index = 0
    for character in lower_text:
        if query_index >= len(lower_query):
            break
        if character == lower_query[query_index]:
            query_index += 1
    return query_index == len(lower_query)


def literal_needle(query: str) -> str:
    """Return the query without its optional wrapping quotes."""
    if query and query[0] in ('"', "'"):
        quote = query[0]
        needle = query[1:]
        if needle.endswith(quote):
            needle = needle[:-1]
        return needle
    return query


def match_literal_query(query: str, text: str) -> bool:
    """Match the whole query as one case-insensitive substring."""
    needle = literal_needle(query)
    if not needle:
        return True
    return needle.lower() in text.lower()


def match_all_terms_query(query: str, text: str) -> bool:
    """Require every whitespace-separated term as a case-insensitive substring.

    For fields where the subsequence grammar is meaningless — a long message
    body, where a short query occurs in order almost every time. Terms match
    anywhere and in any order, so ``release plan`` finds a text that says
    "planning the release".

    A leading quote switches the whole query to one exact phrase, as everywhere
    else in the shared grammar. Quoting is all-or-nothing here: a phrase is
    never one term among others.
    """
    if query and query[0] in ('"', "'"):
        return match_literal_query(query, text)
    lower_text = text.lower()
    return all(term.lower() in lower_text for term in query.split())


def match_text_query(query: str, text: str) -> bool:
    """Apply the shared fuzzy or leading-quote literal filter grammar."""
    if query and query[0] in ('"', "'"):
        return match_literal_query(query, text)
    return match_subsequence(query, text)

File: frontend/src/utils/textFilter.js
/**
 * Shared sidebar filter matching, used by the session list and the artifact
 * bookmark list.
 *
 * - `matchSubsequence`: case-insensitive subsequence (fuzzy) match — every
 *   character of the query appears in the text, in order, not necessarily
 *   consecutive.
 * - `matchQuery`: the dispatch used by sidebar filters. A query starting with
 *   `"` or `'` switches to case-insensitive literal-substring matching (an
 *   optional trailing matching quote is stripped, so both `"foo` and `"foo"`
 *   look for the substring `foo`); anything else uses `matchSubsequence`.
 *
 * The backend mirrors `matchQuery` in `match_text_query`
 * (`twicc/core/text_filter.py`) so the bulk-archive scope matches the sidebar
 * exactly. The peer inbox deliberately does NOT use this grammar: it searches
 * message bodies, where a subsequence match accepts nearly any short query. It
 * calls `match_all_terms_query` instead — every word must appear as a
 * substring, in any order.
 */

/**
 * Case-insensitive subsequence match.
 * @param {string} query - The search query
 * @param {string} text - The text to search in
 * @returns {boolean} True if query is a subsequence of text
 */
export function matchSubsequence(query, text) {
    const lowerQuery = query.toLowerCase()
    const lowerText = text.toLowerCase()

    let queryIndex = 0
    for (let i = 0; i < lowerText.length && queryIndex < lowerQuery.length; i++) {
        if (lowerText[i] === lowerQuery[queryIndex]) {
            queryIndex++
        }
    }
    return queryIndex === lowerQuery.length
}

/**
 * Resolve a sidebar filter query against a display string (see file header).
 * @param {string} query
 * @param {string} text
 * @returns {boolean}
 */
export function matchQuery(query, text) {
    const first = query[0]
    if (first === '"' || first === "'") {
        let needle = query.slice(1)
        if (needle.endsWith(first)) needle = needle.slice(0, -1)
        if (!needle) return true
        return text.toLowerCase().includes(needle.toLowerCase())
    }
    return matchSubsequence(query, text)
}
